storm-client/src/jvm/org/apache/storm/executor/spout/SpoutOutputCollectorImpl.java

```diff
@@ -44,14 +48,15 @@
     private final Boolean isEventLoggers;
     private final Boolean isDebug;
     private final RotatingMap<Long, TupleInfo> pending;
+    private TupleInfo globalTupleInfo = new TupleInfo();  // thread safety: assumes Collector.emit*() calls are externally synchronized (if needed).
```

Comment:
Accessing the Thread Local (TL) instance via ThreadLocal.get() typically involves a map lookup behind the scenes.

**Related Note:** I reluctantly used TL for latching on to JCQueue.BatchInserter instance for the producers to JCQueue. Reluctant since I noticed perf hit when doing some targeted microbenchmarking. I used it anyway because it was a perf improvement over the ConcurrentHashMap employed in Disruptor, and eliminating TL needed a bigger change to the interface and producers. I think it is possible to achieve TL free JCQueue and gain some perf.. perhaps in a follow up jira.

Although many decisions were measured, due to scope it was not feasible to measure each one. So, in the critical path, I have taken this general approach of :
- avoid locks & synchronization ... and try using lock/wait-free approaches where synchronization is unavoidable.
- avoid map lookups and object creation

This was a case of avoiding synchronization, (TL) map lookups & object allocation.

Reply:
Can we do a sanity check for the fast path + documentation?  Check if the thread id is the same as the id of the main thread we expect emits to come from.  If so we go with the fast path, if not we have a thread local or do some kind of locking + documentation about why you never want the spout to emit from a background thread.

Reply:
To get the id of the current thread  involves a call to Thread.currentThread() which is quite [expensive](!http://www.jutils.com/checks/performance.html)... so not good to use to determine whether or not to use fast path.

I am introducing that check if topology.debug is enabled as a compromise. This mode could 
 be used mode to do any checks in dev mode that are unnecessary or expensive to do repeatedly in production.

I have opened: [STORM-2945](!https://issues.apache.org/jira/browse/STORM-2945) to nail down and document background emits support.. we can document both spout & bolt support semantics together in the same document.